enterprise/tailnet/pgcoord.go

```diff
@@ -683,21 +899,116 @@ func (q *querier) newConn(c *connIO) {
 		return
 	}
 	cm.count++
-	q.conns[c] = struct{}{}
-	go q.cleanupConn(c)
+	q.conns[c.UniqueID()] = c
+	go q.waitCleanupConn(c)
 }
 
-func (q *querier) cleanupConn(c *connIO) {
-	<-c.ctx.Done()
+func (q *querier) newClientSubscription(c agpl.Queue, agentID uuid.UUID) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
-	delete(q.conns, c)
+
+	if _, ok := q.clientSubscriptions[c.UniqueID()]; !ok {
+		q.clientSubscriptions[c.UniqueID()] = map[uuid.UUID]struct{}{}
+	}
+
 	mk := mKey{
-		agent: c.agent,
-		// if client is Nil, this is an agent connection, and it wants the mappings for all the clients of itself
-		clientsOfAgent: c.client == uuid.Nil,
+		agent: agentID,
+		kind:  agpl.QueueKindClient,
+	}
+	cm, ok := q.mappers[mk]
+	if !ok {
+		ctx, cancel := context.WithCancel(q.ctx)
+		mpr := newMapper(ctx, q.logger, mk, q.heartbeats)
+		cm = &countedMapper{
+			mapper: mpr,
+			count:  0,
+			cancel: cancel,
+		}
+		q.mappers[mk] = cm
+		// we don't have any mapping state for this key yet
+		q.workQ.enqueue(mk)
+	}
+	if err := sendCtx(cm.ctx, cm.add, c); err != nil {
+		return
+	}
+	q.clientSubscriptions[c.UniqueID()][agentID] = struct{}{}
+	cm.count++
+}
+
+func (q *querier) removeClientSubscription(c agpl.Queue, agentID uuid.UUID) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	mk := mKey{
+		agent: agentID,
+		kind:  agpl.QueueKindClient,
 	}
 	cm := q.mappers[mk]
+	if err := sendCtx(cm.ctx, cm.del, c); err != nil {
+		return
+	}
+	delete(q.clientSubscriptions[c.UniqueID()], agentID)
+	cm.count--
+	if cm.count == 0 {
+		cm.cancel()
+		delete(q.mappers, mk)
+	}
+}
+
+func (q *querier) newClientConn(c agpl.Queue) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if !q.healthy {
+		err := c.Close()
+		q.logger.Info(q.ctx, "closed incoming connection while unhealthy",
+			slog.Error(err),
+			slog.F("client_id", c.UniqueID()),
+		)
+		return
+	}
+
+	q.conns[c.UniqueID()] = c
+	go q.waitCleanupConn(c)
+}
+
+func (q *querier) waitCleanupConn(c agpl.Queue) {
+	<-c.Done()
+	q.cleanupConn(c)
+}
+
+func (q *querier) cleanupConn(c agpl.Queue) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	delete(q.conns, c.UniqueID())
+
+	// Iterate over all subscriptions and remove them from the mappers.
+	for agentID := range q.clientSubscriptions[c.UniqueID()] {
+		mk := mKey{
+			agent: agentID,
+			kind:  c.Kind(),
+		}
+		cm, ok := q.mappers[mk]
+		if ok {
+			if err := sendCtx(cm.ctx, cm.del, c); err != nil {
+				continue
+			}
+			cm.count--
+			if cm.count == 0 {
+				cm.cancel()
+				delete(q.mappers, mk)
+			}
+		}
+	}
```

Comment:
this should delete from the `q.clientSubscriptions` map too --- otherwise you remove from the mappers but leave the subscriptions in the map.  c.f. comment on line 991.